tests/string_test.c

```diff
@@ -0,0 +1,310 @@
+    expected = "";
+    ASSERT_BIN_ARRAYS_EQUALS(expected, strlen(expected), value.ptr,  value.len);
+
+    ASSERT_SUCCESS(aws_array_list_get_at(&output, &value, 2));
+    expected = "testb";
+    ASSERT_BIN_ARRAYS_EQUALS(expected, strlen(expected), value.ptr,  value.len);
+
+    ASSERT_SUCCESS(aws_array_list_get_at(&output, &value, 3));
+    expected = "testc";
+    ASSERT_BIN_ARRAYS_EQUALS(expected, strlen(expected), value.ptr,  value.len);
+
+    aws_array_list_clean_up(&output);
+
+    return 0;
+}
+
+AWS_TEST_CASE(test_char_split_adj_tokens, test_char_split_adj_tokens_fn)
+
+static int test_char_split_with_max_splits_fn(struct aws_allocator *allocator, void *ctx) {
+    const char str_to_split[] = ";testa;testb;testc";
+
+    struct aws_byte_buf to_split = aws_byte_buf_from_literal(str_to_split);
+
+    struct aws_array_list output;
+    ASSERT_SUCCESS(aws_array_list_init_dynamic(&output, allocator, 4, sizeof(struct aws_byte_cursor)));
+    ASSERT_SUCCESS(aws_string_split_on_char_n(&to_split, ';', &output, 2));
+    ASSERT_INT_EQUALS(3, aws_array_list_length(&output));
+
+    struct aws_byte_cursor value = { 0 };
+
+    ASSERT_SUCCESS(aws_array_list_get_at(&output, &value, 0));
+    char *expected = "";
+    ASSERT_BIN_ARRAYS_EQUALS(expected, strlen(expected), value.ptr, value.len);
+
+    ASSERT_SUCCESS(aws_array_list_get_at(&output, &value, 1));
+    expected = "testa";
+
+    ASSERT_BIN_ARRAYS_EQUALS(expected, strlen(expected), value.ptr, value.len);
+
+    ASSERT_SUCCESS(aws_array_list_get_at(&output, &value, 2));
+    expected = "testb;testc";
+    ASSERT_BIN_ARRAYS_EQUALS(expected, strlen(expected), value.ptr, value.len);
+
+    aws_array_list_clean_up(&output);
+
+    return 0;
+}
+
+AWS_TEST_CASE(test_char_split_with_max_splits, test_char_split_with_max_splits_fn)
+
+static int test_char_split_output_too_small_fn(struct aws_allocator *allocator, void *ctx) {
+    const char str_to_split[] = "testa;testb;testc;";
+
+    struct aws_byte_buf to_split = aws_byte_buf_from_literal(str_to_split);
+
+    struct aws_array_list output;
+    struct aws_byte_buf output_array[3] = {{0}};
+    ASSERT_SUCCESS(aws_array_list_init_static(&output, output_array, 3, sizeof(struct aws_byte_cursor)));
+    ASSERT_ERROR(AWS_ERROR_LIST_EXCEEDS_MAX_SIZE, aws_string_split_on_char(&to_split, ';', &output));
+   
+    return 0;
+}
+
+AWS_TEST_CASE(test_char_split_output_too_small, test_char_split_output_too_small_fn)
+
+static int test_buffer_cat_fn(struct aws_allocator *allocator, void *ctx) {
+    struct aws_byte_buf str1 = aws_byte_buf_from_literal("testa");
+    struct aws_byte_buf str2 = aws_byte_buf_from_literal(";testb");
+    struct aws_byte_buf str3 = aws_byte_buf_from_literal(";testc");
+
+
+    const char expected[] = "testa;testb;testc";
+
+    struct aws_byte_buf destination;
+    ASSERT_SUCCESS(aws_byte_buf_alloc(allocator, &destination, str1.len + str2.len + str3.len));
+    ASSERT_SUCCESS(aws_byte_buf_cat(&destination, 3, &str1, &str2, &str3));
```

Comment:
Given that we have to manually compute the sum of the input buffers here, would it make more sense to move the allocation into cat to make it more convenient to use?

Reply:
I’d rather not make the allocation decision for the user here. For example, I could already have preallocated buffers I want to use.